test/python/transpiler/test_optimize_annotated.py

```diff
@@ -193,3 +196,45 @@ def test_if_else(self):
         )
 
         self.assertEqual(qc_optimized, expected_qc)
+
+    def test_conjugate_reduction(self):
+        """Test conjugate reduction optimization."""
+
+        # Create a control-annotated operation;
+        # the definition of the base operation has conjugate decomposition P -- Q -- R with R = P^{-1}
+        qc_def = QuantumCircuit(6)
+        qc_def.cx(0, 1)  # P
+        qc_def.z(0)  # P
+        qc_def.s(0)  # P
+        qc_def.cx(0, 4)  # P
+        qc_def.cx(4, 3)  # P
+        qc_def.y(3)  # Q
+        qc_def.cx(3, 0)  # Q
+        qc_def.cx(4, 3)  # R
+        qc_def.cx(0, 4)  # R
+        qc_def.sdg(0)  # R
+        qc_def.z(0)  # R
+        qc_def.cx(0, 1)  # R
+        qc_def.z(5)  # P
+        qc_def.z(5)  # R
+        qc_def.x(2)  # Q
+        custom = qc_def.to_gate().control(annotated=True)
+
+        # Create a quantum circuit with an annotated operation.
+        qc = QuantumCircuit(8)
+        qc.cx(0, 2)
+        qc.append(custom, [0, 1, 3, 4, 5, 7, 6])
+        qc.h(0)
+        qc.z(4)
+
+        qc_keys = qc.count_ops().keys()
+        self.assertIn("annotated", qc_keys)
+
+        # Run optimization pass.
+        qc_optimized = OptimizeAnnotated(basis_gates=["cx", "u"])(qc)
+
+        # THe pass should simplify the gate
+        qc_optimized_keys = qc_optimized.count_ops().keys()
+        self.assertIn("optimized", qc_optimized_keys)
+        self.assertNotIn("annotated", qc_optimized_keys)
+        self.assertEqual(Operator(qc), Operator(qc_optimized))
```

Comment:
It would be good to make some assertions that the pass has collected the gates into the right parts of the decomposition.

Reply:
Thanks. I have added multiple tests exploring various collection cases.